src/dune_pkg/archive_type.ml

```diff
@@ -2,18 +2,45 @@ open Stdune
 module Process = Dune_engine.Process
 open Fiber.O
 
-let bin =
-  lazy
-    (match Bin.which ~path:(Env_path.path Env.initial) "tar" with
-     | Some x -> x
-     | None -> Dune_engine.Utils.program_not_found "tar" ~loc:None)
+type t =
+  { bin : Path.t Lazy.t
+  ; suffixes : string list
+    (* Identify a file by its suffix rather than by its extension, since
+       ".tar.gz" files only have an extension of ".gz". *)
+  ; build_args : archive:Path.t -> target_in_temp:Path.t -> string list
+  }
+
+let make ~bin_name ~suffixes ~build_args =
+  let bin =
+    lazy
+      (match Bin.which ~path:(Env_path.path Env.initial) bin_name with
+       | Some x -> x
+       | None -> Dune_engine.Utils.program_not_found bin_name ~loc:None)
+  in
+  { bin; suffixes; build_args }
+;;
+
+let tar =
+  make
+    ~bin_name:"tar"
+    ~suffixes:[ ".tar"; ".tar.gz"; ".tgz"; ".tar.bz2"; ".tbz" ]
+    ~build_args:(fun ~archive ~target_in_temp ->
+      [ "xf"; Path.to_string archive; "-C"; Path.to_string target_in_temp ])
+;;
+
+let zip =
+  make ~bin_name:"unzip" ~suffixes:[ ".zip" ] ~build_args:(fun ~archive ~target_in_temp ->
+    [ Path.to_string archive; "-d"; Path.to_string target_in_temp ])
```

Comment:
Does that mean that to run it we need both `tar` and `unzip` in the path, even if we never use e.g. `unzip`?

Maybe it would be better to conditionally extend the list of supported suffixes depending on what binaries we do find and only fail if we encounter a suffix we don't support? That way Dune distribution packagers don't have to declare a (non-optional) dependency on `unzip`.

Reply:
Looking up the named binary in PATH is lazy and only forced in `extract`, so systems without `unzip` will only fail when trying to unzip a .zip archive.

Reply:
I double-checked this by building a project with tar.gz dependencies only on a docker container without `unzip` installed.